docs/ldml/tr35-keyboards.md

```diff
@@ -2030,6 +2043,316 @@ Footnotes:
+
+> <small>
+>
+> Parents: [test](#Element_test)
+>
+> Children: _none_
+>
+> Occurrence: Optional, Single
+> </small>
+
+_Attribute:_ `to` (required)
+
+Specifies the starting context. This text may be escaped with `\u` notation, see [Escaping](#Escaping).
+
+**Example**
+
+```xml
+<startContext to="abc\u0022"/>
+```
+
+
+### Test Element: event
+
+> <small>
+>
+> Parents: [test](#test-element-test)
+>
+> Children: _none_
+>
+> Occurrence: Optional, Multiple
+> </small>
+
+This element represents a single keystroke or other gesture event.
+
+Optionally, one of the gesture attributes, either `flick`, `longPress`, or `tapCount` may be specified. If none of the gesture attributes are specified, then a regular keypress is effected on the key.  It is an error to specify more than one gesture attribute.
+
+If a key is not found, or a particular gesture has no definition, the output should be behave as if the user attempted to perform such an action.  For example, an unspecified flick would result in no output.
+
+_Attribute:_ `key` (required)
+
+This attribute specifies a key by means of the key’s `id` attribute.
+
+_Attribute:_ `flick`
+
+This attribute specifies a flick gesture to be performed on the specified key instead of a keypress, such as `e` or `nw se`. This value corresponds to the `directions` attribute of the [`<flick>`](#Element_flicks) element.
+
+_Attribute:_ `longPress`
+
+This attribute specifies that a long press gesture should be performed on the specified key instead of a keypress. For example, `longPress="2"` indicates that the second character in a longpress series should be chosen. `longPress="0"` indicates that the `longPressDefault` value, if any, should be chosen. This corresponds to `longPress` and `longPressDefault` on [`<key>`](#Element_key).
+
+_Attribute:_ `tapCount`
+
+This attribute specifies that a multi-tap gesture should be performed on the specified key instead of a keypress. For example, `tapCount="3"` indicates that the key should be tapped three times in rapid succession. This corresponds to `multiTap` on [`<key>`](#Element_key).
+
+Note that `tapCount="1"` is valid, but represents an ordinary keypress.
+
+**Example**
+
+```xml
+<keystroke key="q"/>
+<keystroke key="doublequote"/>
+<keystroke key="s" flick="nw se"/>
+<keystroke key="e" longPress="1"/>
+<keystroke key="E" tapCount="2"/>
+```
+
+### Test Element: emit
+
+> <small>
+>
+> Parents: [test](#test-element-emit)
+>
+> Children: _none_
+>
+> Occurrence: Optional, Multiple
+> </small>
+
+This element represents a text output event.
+
+_Attribute:_ `to` (required)
+
+This attribute specifies a a string of output text, which is intended to match a key’s `to` attribute.
+
+Tooling should give a warning if this attribute matches more than one key, or does not match any keys.
```

Comment:
Be good to note that this is helpful for testing transforms without needing reference to the key definitions.

Don't think we should be generating a warning if more than one key generates the same `to`. This will be common, for example, with punctuation on touch layouts. A warning if the `to` cannot be generated by any key/flick/longPress/multiTap is appropriate.

Reply:
To be clear the warning isn't that keys match more than one to, but that such a key is used in the test. Better to use keystroke to be un ambiguous.

Reply:
I don't think so: when we want to write tests for transforms, we probably don't really care which key is generating the character(s); this would only be a concern if no key could generate the character.

Realised just now that we need to clarify that `<emit>` will inject `to` into the transform+reorder engine for the keyboard -- that it isn't emitted directly into the text store.

Reply:
Good point. Since flicks and such are out of the picture, it doesn't matter.